**Penjelasan Dataset (Dataset Summary)**



Dataset ini berisi informasi demografis, sosioekonomi,
dan akademik dari mahasiswa.

---
**Penjelasan setiap kolom:**
* **`Unnamed: 0`**: Nomor indeks baris data yang dibuat otomatis sebagai penunjuk setiap record data saja (bukan UID).
* **`Student Id`**: Nomor identifikasi unik yang diberikan kepada setiap mahasiswa.
* **`Marital status`**: Status perkawinan mahasiswa
* **`Application mode`**: Kode numerik yang mewakili jalur pendaftaran mahasiswa
* **`Application order`**: Urutan prioritas pilihan jurusan yang diambil. Angka yang lebih kecil berarti jurusan tersebut adalah pilihan utama.
* **`Course`**: Kode khusus untuk program studi spesifik yang diambil mahasiswa.
* **`Daytime/evening attendance`**: Menunjukkan waktu perkuliahan mahasiswa
* **`Previous qualification`**: Kode untuk tingkat pendidikan terakhir mahasiswa sebelum masuk universitas
* **`Previous qualification (grade)`**: Nilai akhir dari tingkat pendidikan sebelumnya.
* **`Nacionality`**: Kewarganegaraan mahasiswa.
* **`Mother's qualification`** & **`Father's qualification`**: Kode untuk tingkat pendidikan tertinggi yang dimiliki oleh ibu dan ayah.
* **`Mother's occupation`** & **`Father's occupation`**: Kode untuk jenis pekerjaan ibu dan ayah.
* **`Admission grade`**: Nilai akhir yang diperoleh mahasiswa saat seleksi masuk universitas.
* **`Displaced`**: Status perpindahan mahasiswa untuk kuliah apakah pergi dari daerah asal atau tidak
* **`Educational special needs`**: Indikator apakah mahasiswa memiliki kebutuhan pendidikan khusus.
* **`Debtor`**: Status tunggakan utang mahasiswa kepada universitas.
* **`Tuition fees up to date`**: Status pembayaran biaya kuliah.
* **`Gender`**: Jenis kelamin mahasiswa.
* **`Scholarship holder`**: Status sebagai penerima beasiswa.
* **`Age at enrollment`**: Usia mahasiswa saat pertama kali mendaftar.
* **`International`**: Status jika tergolong mahasiswa internasional atau tidak.
* **`Curricular units (credited)`**: Jumlah SKS/mata kuliah yang diakui dari pendidikan sebelumnya.
* **`Curricular units (enrolled)`**: Jumlah total SKS/mata kuliah yang diambil pada semester tersebut.
* **`Curricular units (evaluations)`**: Jumlah ujian/evaluasi yang diikuti oleh mahasiswa pada semester tersebut.
* **`Curricular units (approved)`**: Jumlah SKS/mata kuliah yang disetujui pada semester tersebut.
* **`Curricular units (grade)`**: Nilai rata-rata dari semua mata kuliah yang dievaluasi di semester itu.
* **`Curricular units (without evaluations)`**: Jumlah mata kuliah yang terdaftar namun tidak diikuti ujian/evaluasinya.
* **`Unemployment rate`**: Tingkat pengangguran di wilayah tersebut.
* **`Inflation rate`**: Tingkat inflasi di wilayah tersebut.
* **`GDP` (Gross Domestic Product)**: Produk Domestik Bruto, sebuah indikator pertumbuhan ekonomi wilayah/negara.
* **`Target`**: Variabel dependen yang menjadi tujuan prediksi nantinya dan untuk klasifikasi. Kolom ini berisi status akhir dari setiap mahasiswa (`Graduate`, `Dropout`, `Enrolled`)

###IMPORTING LIBRARY

In [1]:
!pip install miceforest==5.7.0 numpy==1.26.4 lightgbm==3.3.5
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
import miceforest as mf
import re

###**EDA**

###1. **Show raw datasets**

In [2]:
# Inisialization
train_features = pd.read_csv('features.csv')
train_labels = pd.read_csv('labels.csv')

# Menampilkan dataset features.csv
train_features

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 1st sem (without evaluations)  \
0                                                0.0   
1                                                0.0   
2                                                0.0   
3                                                0.0   
4                                                0.0   
...                                              ...   
3091                                             0.0   
3092                                             0.0   
3093                                             0.0   
3094                                             0.0   
3095                                             NaN   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   

In [3]:
# Menampilkan dataset labels.csv
train_labels

Target
0     Graduate
1     Graduate
2     Graduate
3     Graduate
4      Dropout
...        ...
3091  Graduate
3092  Graduate
3093  Graduate
3094   Dropout
3095  Graduate

[3096 rows x 1 columns]

###2. **Datasets aggregation**

In [4]:
# Proses agregasi antara train_features dan train_labels melalui kolom (axis = 1)
df_mhs = pd.concat([train_features, train_labels], axis = 1)

# Menampilkan dataframe setelah diagregasi
df_mhs

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

###3. **Show data types**

In [5]:
# Menampilkan tipe data setiap kolom pada dataframe
df_mhs.dtypes

Unnamed: 0                                          int64
Student Id                                          int64
Marital status                                     object
Application mode                                  float64
Application order                                 float64
Course                                            float64
Daytime/evening attendance\t                       object
Previous qualification                            float64
Previous qualification (grade)                    float64
Nacionality                                        object
Mother's qualification                            float64
Father's qualification                            float64
Mother's occupation                               float64
Father's occupation                               float64
Admission grade                                   float64
Displaced                                         float64
Educational special needs                         float64
Debtor                                            float64
Tuition fees up to date                           float64
Gender                                            float64
Scholarship holder                                float64
Age at enrollment                                 float64
International                                     float64
Curricular units 1st sem (credited)               float64
Curricular units 1st sem (enrolled)               float64
Curricular units 1st sem (evaluations)            float64
Curricular units 1st sem (approved)               float64
Curricular units 1st sem (grade)                  float64
Curricular units 1st sem (without evaluations)    float64
Curricular units 2nd sem (credited)               float64
Curricular units 2nd sem (enrolled)               float64
Curricular units 2nd sem (evaluations)            float64
Curricular units 2nd sem (approved)               float64
Curricular units 2nd sem (grade)                  float64
Curricular units 2nd sem (without evaluations)    float64
Unemployment rate                                 float64
Inflation rate                                    float64
GDP                                               float64
Target                                             object
dtype: object

###4. **Descriptive statistic and analytics**

In [6]:
# Menampilkan statistika deskriptif dari dataframe
df_mhs.describe()

Unnamed: 0   Student Id  Application mode  Application order  \
count  3096.000000  3096.000000       2924.000000        2988.000000   
mean   1547.500000  5470.623062         18.785568           1.734940   
std     893.882543  2582.483019         17.532674           1.327527   
min       0.000000  1000.000000          1.000000           1.000000   
25%     773.750000  3288.500000          1.000000           1.000000   
50%    1547.500000  5441.500000         17.000000           1.000000   
75%    2321.250000  7712.250000         39.000000           2.000000   
max    3095.000000  9997.000000         53.000000           9.000000   

            Course  Previous qualification  Previous qualification (grade)  \
count  2944.000000             2973.000000                     3011.000000   
mean   8817.929348                4.685503                      132.647293   
std    2148.407432               10.439709                       13.322662   
min      33.000000                1.000000                       95.000000   
25%    9085.000000                1.000000                      124.000000   
50%    9238.000000                1.000000                      133.100000   
75%    9556.000000                1.000000                      140.000000   
max    9991.000000               43.000000                      190.000000   

       Mother's qualification  Father's qualification  Mother's occupation  \
count             2951.000000             3000.000000          3034.000000   
mean                19.470010               22.408667            11.495715   
std                 15.636952               15.291008            28.173386   
min                  1.000000                1.000000             0.000000   
25%                  2.000000                3.000000             4.000000   
50%                 19.000000               19.000000             5.000000   
75%                 37.000000               37.000000             9.000000   
max                 44.000000               44.000000           194.000000   

       ...  Curricular units 1st sem (without evaluations)  \
count  ...                                     3023.000000   
mean   ...                                        0.133311   
std    ...                                        0.666707   
min    ...                                        0.000000   
25%    ...                                        0.000000   
50%    ...                                        0.000000   
75%    ...                                        0.000000   
max    ...                                       10.000000   

       Curricular units 2nd sem (credited)  \
count                          3035.000000   
mean                              0.516639   
std                               1.840537   
min                               0.000000   
25%                               0.000000   
50%                               0.000000   
75%                               0.000000   
max                              19.000000   

       Curricular units 2nd sem (enrolled)  \
count                          2997.000000   
mean                              6.208876   
std                               2.185127   
min                               0.000000   
25%                               5.000000   
50%                               6.000000   
75%                               7.000000   
max                              23.000000   

       Curricular units 2nd sem (evaluations)  \
count                             2889.000000   
mean                                 8.061613   
std                                  3.942850   
min                                  0.000000   
25%                                  6.000000   
50%                                  8.000000   
75%                                 10.000000   
max                                 27.000000   

       Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
count                          2945.000000

**Analisis singkat statistik deskriptif dataset**

---

1. Kualitas Data dan Data yang Hilang (Missing Values)

* **Jumlah Data Tidak Konsisten:** Total entri dalam dataset adalah 3096, namun hampir semua kolom lain memiliki `count` yang lebih rendah.
    * **Contoh:** Kolom `Application mode` hanya memiliki 2924 data, sedangkan `Curricular units 2nd sem (evaluations)` hanya 2889 data. Dari sini menunjukkan adanya *missing values* secara signifikan di seluruh dataset.

2. Profil Pendaftaran Mahasiswa

* **Prioritas Jurusan (`Application order`):**
    * *Rata-rata (`mean`):* 1.73
    * *Median (`50%`):* 1.0
    * *Kuartil Atas (`75%`):* 2.0
    * **Wawasan:** Setengah dari mahasiswa (`50%`) masuk ke jurusan yang merupakan **pilihan pertama** mereka. Bahkan, 75% mahasiswa masuk ke jurusan pilihan pertama atau kedua. Ini menandakan mayoritas mahasiswa berada di program studi yang mereka inginkan.

* **Nilai Kualifikasi Sebelumnya (`Previous qualification (grade)`):**
    * *Rata-rata (`mean`):* 132.6
    * *Rentang (Min-Max):* 95 hingga 190
    * **Wawasan:** Terdapat rentang nilai yang cukup lebar dari calon mahasiswa yang diterima, menunjukkan variasi dalam latar belakang akademik mereka.

* **Kolom Kategorikal:**
    * **Peringatan:** Untuk kolom yang nilainya adalah kode kategori (seperti `Course`, `Mother's qualification`), statistik seperti `mean` (rata-rata) tidak memiliki interpretasi yang relevan karena angka tersebut hanyalah sebuah label.

3. Analisis Performa Akademik

Mari kita fokus pada data semester kedua sebagai contoh performa mahasiswa.

* **Beban Studi (`Curricular units 2nd sem (enrolled)`):**
    * *Rata-rata (`mean`):* 6.2 mata kuliah
    * *Median (`50%`):* 6.0 mata kuliah
    * **Wawasan:** Rata-rata mahasiswa mengambil sekitar 6 mata kuliah per semester, dengan distribusi yang cukup normal.

* **Tingkat Kelulusan Mata Kuliah (`Curricular units 2nd sem (approved)`):**
    * *Rata-rata (`mean`):* 4.4 mata kuliah
    * *Median (`50%`):* 5.0 mata kuliah
    * *Kuartil Bawah (`25%`):* 2.0 mata kuliah
    * **Wawasan:** Rata-rata, mahasiswa hanya lulus sekitar 4.4 dari 6.2 mata kuliah yang diambil. Ini adalah **indikator kuat adanya tantangan akademik**. Fakta bahwa 25% mahasiswa hanya lulus 2 mata kuliah atau kurang menunjukkan ada segmen mahasiswa yang benar-benar kesulitan.

* **Nilai Akademik (`Curricular units 2nd sem (grade)`):**
    * *Rata-rata (`mean`):* 10.27
    * *Median (`50%`):* 12.25
    * **Wawasan:** Median lebih tinggi dari rata-rata. Ini mengindikasikan **distribusi nilai yang condong ke kiri (*left-skewed*)**, artinya ada sejumlah mahasiswa dengan nilai sangat rendah yang "menarik" nilai rata-rata ke bawah. Setengah dari mahasiswa sebenarnya memiliki nilai di atas 12.25.

4. Konteks Makroekonomi

* **Rentang Data:**
    * `Unemployment rate`: Berkisar antara 7.6% hingga 16.2%.
    * `Inflation rate`: Berkisar antara -0.8% (deflasi) hingga 3.7%.
    * `GDP`: Berkisar antara -4.06% (resesi) hingga 3.51% (pertumbuhan).
    * **Wawasan:** Dataset ini mencakup periode waktu dengan kondisi ekonomi yang sangat bervariasi (baik dan buruk). Variasi ini bisa menjadi fitur yang sangat berguna untuk model prediksi.
---
Ringkasan Poin Kunci (Key Takeaways)

1.  **Masalah Data Hilang:** Prioritas utama adalah menangani *missing values* yang signifikan di banyak kolom.
2.  **Mahasiswa Termotivasi:** Sebagian besar mahasiswa belajar di jurusan pilihan utama mereka.
3.  **Tantangan Akademik Nyata:** Ada kesenjangan yang jelas antara jumlah mata kuliah yang diambil dan yang dilulusi, menandakan area risiko bagi mahasiswa.
4.  **Distribusi Nilai Tidak Normal:** Adanya populasi mahasiswa berisiko dengan nilai sangat rendah mempengaruhi performa akademik rata-rata secara keseluruhan.
5.  **Konteks Ekonomi Bervariasi:** Dataset ini kaya akan variasi data ekonomi yang dapat dimanfaatkan untuk analisis lebih lanjut.

###5. **Slicing data**

In [7]:
# Menampilkan 20 baris data pertama pada dataframe
df_mhs.head(20)

Unnamed: 0  Student Id Marital status  Application mode  \
0            0        6634         single               1.0   
1            1        6459         single               1.0   
2            2        2238         single              42.0   
3            3        2479         single               1.0   
4            4        4260         single              44.0   
5            5        3544        married              39.0   
6            6        4956         single              39.0   
7            7        6758         single               1.0   
8            8        2484         single               1.0   
9            9        9845         single              39.0   
10          10        3690         single               1.0   
11          11        2585         single               1.0   
12          12        5241         single               1.0   
13          13        2972         single               1.0   
14          14        2003        married              39.0   
15          15        3007         single               7.0   
16          16        9857         single              17.0   
17          17        3327         single               NaN   
18          18        5235         single              17.0   
19          19        3969         single               1.0   

    Application order  Course Daytime/evening attendance\t  \
0                 3.0  9773.0                          day   
1                 1.0     NaN                          day   
2                 1.0  9500.0                          day   
3                 1.0     NaN                          day   
4                 1.0     NaN                          day   
5                 1.0  9853.0                          day   
6                 1.0  9085.0                          day   
7                 4.0  9670.0                          day   
8                 1.0  9773.0                          day   
9                 1.0  9670.0                          day   
10                1.0  9670.0                          day   
11                1.0  9070.0                          day   
12                1.0   171.0                          day   
13                3.0  9500.0                          day   
14                1.0  8014.0                          NaN   
15                1.0  9119.0                          day   
16                1.0  9119.0                          day   
17                1.0  9003.0                          day   
18                1.0   171.0                          day   
19                1.0  9085.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                      1.0                             NaN  Portuguese  ...   
1                      1.0                           136.0  Portuguese  ...   
2                      1.0                           120.0  Portuguese  ...   
3                      1.0                           141.0  Portuguese  ...   
4                     39.0                           120.0     Spanish  ...   
5                      1.0                             NaN  Portuguese  ...   
6                      1.0                           133.1         NaN  ...   
7                      1.0                           130.0  Portuguese  ...   
8                      1.0                           135.0  Portuguese  ...   
9                      1.0                           120.0  Portuguese  ...   
10                     1.0                           133.1  Portuguese  ...   
11                     1.0                           117.0  Portuguese  ...   
12                     1.0                           162.0  Portuguese  ...   
13                     1.0                           135.0  Portuguese  ...   
14                     NaN                           120.0  Portuguese  ...   
15                     1.0                           140.8  Portuguese  ...   
16                     1.0        

In [8]:
# Menampilkan 20 baris data terakhir pada dataframe
df_mhs.tail(20)

Unnamed: 0  Student Id Marital status  Application mode  \
3076        3076        6757         single              17.0   
3077        3077        5750         single               1.0   
3078        3078        2599         single               1.0   
3079        3079        3249         single              17.0   
3080        3080        3366       divorced              39.0   
3081        3081        3462         single              39.0   
3082        3082        9194         single              17.0   
3083        3083        5574         single               1.0   
3084        3084        8311         single               1.0   
3085        3085        2304         single              17.0   
3086        3086        6615         single              43.0   
3087        3087        3759         single               1.0   
3088        3088        8085         single              51.0   
3089        3089        1573         single              17.0   
3090        3090        9737         single              51.0   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
3076                NaN  9254.0                          day   
3077                2.0  9500.0                          day   
3078                1.0  9085.0                          day   
3079                1.0  9238.0                          day   
3080                1.0  9003.0                          day   
3081                1.0   171.0                          day   
3082                5.0  9147.0                          day   
3083                1.0  9500.0                          day   
3084                2.0  9773.0                          day   
3085                1.0  9238.0                          day   
3086                1.0  9556.0                          day   
3087                1.0   171.0                          day   
3088                1.0  9670.0                          day   
3089                2.0   171.0                          day   
3090                1.0  9070.0                          day   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
3076                     1.0                           122.0  Portuguese  ...   
3077                     1.0                           140.0  Portuguese  ...   
3078                     1.0                           131.0  Portuguese  ...   
3079                     1.0                           120.0  Portuguese  ...   
3080                    12.0                           133.1  Portuguese  ...   
3081                     1.0                           147.0  Portuguese  ...   
3082                     1.0                           131.0  Portuguese  ...   
3083                     1.0                           140.0  Portuguese  ...   
3084                     1.0                           130.0  Portuguese  ...   
3085                     NaN                           139.0  Portuguese  ...   
3086                     1.0                           149.0  Portuguese  ...   
3087                     3.0                           116.0  Portuguese  ...   
3088                     1.0                           133.1  Portuguese  ...   
3089                     1.0                           142.0  Portuguese  ...   
3090                     1.0                           143.0  Portuguese  ...   

In [9]:
# Menampilkan record data pada dataframe di index 80-100
df_mhs[80:101]

Unnamed: 0  Student Id Marital status  Application mode  \
80           80        1238         single              17.0   
81           81        4121         single              39.0   
82           82        2890         single               1.0   
83           83        4358         single               1.0   
84           84        4896         single               NaN   
85           85        4735         single              18.0   
86           86        1047         single               1.0   
87           87        7430         single              39.0   
88           88        8527         single               1.0   
89           89        5420         single              43.0   
90           90        5625         single              44.0   
91           91        8990            NaN               1.0   
92           92        9143        married              44.0   
93           93        1944        married               7.0   
94           94        3666         single               1.0   
95           95        6960         single              17.0   
96           96        7620         single               1.0   
97           97        6083         single              42.0   
98           98        5525         single              17.0   
99           99        6029         single              17.0   
100         100        9749         single              39.0   

     Application order  Course Daytime/evening attendance\t  \
80                 1.0  9670.0                          day   
81                 1.0  9003.0                          day   
82                 1.0   171.0                          day   
83                 1.0  9147.0                          day   
84                 1.0  9991.0                      evening   
85                 1.0  9500.0                          day   
86                 1.0  9773.0                          day   
87                 2.0  9556.0                          day   
88                 2.0  9500.0                          day   
89                 1.0     NaN                          day   
90                 1.0  9003.0                          day   
91                 1.0  9500.0                          day   
92                 1.0  9085.0                          day   
93                 1.0  9070.0                          day   
94                 3.0  9070.0                          day   
95                 3.0  9119.0                          day   
96                 1.0  9500.0                          day   
97                 1.0  9500.0                          day   
98                 5.0  9238.0                          day   
99                 1.0  9773.0                          day   
100                1.0  9085.0                          day   

     Previous qualification  Previous qualification (grade) Nacionality  ...  \
80                      1.0                           155.0  Portuguese  ...   
81                      1.0                             NaN  Portuguese  ...   
82                      1.0                           159.0         NaN  ...   
83                      1.0                           150.0   Ukrainian  ...   
84                      1.0                           140.0  Portuguese  ...   
85                      1.0                           144.0  Portuguese  ...   
86                      1.0                           117.0  Portuguese  ...   
87                     19.0                           133.1  Portuguese  ...   
88                      1.0                           145.0  Portuguese  ...   
89                      1.0                           121.0  Portuguese  ...   
90                     39.0                           140.0  Portuguese  ...   
91                      1.0                           141.0  Portuguese  ...   
92                     39.0                           160.0  Portuguese  ...   
93                     43.0                           140.0  Portuguese  ...   
94       

###**DATA PREPROCESSING**

###6. **Handling missing values (*encoding & imputing*)**

In [10]:
# Menampilkan dataframe terkini sebelum di encode and impute
df_mhs

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

In [11]:
# Mengidentifikasi kolom yang memiliki null values
df_mhs.isna().sum()

Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                            116
Tuition fees up to date                           156
Gender                                            202
Scholarship holder                                166
Age at enrollment                                 204
International                                     101
Curricular units 1st sem (credited)               208
Curricular units 1st sem (enrolled)               143
Curricular units 1st sem (evaluations)             67
Curricular units 1st sem (approved)               184
Curricular units 1st sem (grade)                   89
Curricular units 1st sem (without evaluations)     73
Curricular units 2nd sem (credited)                61
Curricular units 2nd sem (enrolled)                99
Curricular units 2nd sem (evaluations)            207
Curricular units 2nd sem (approved)               151
Curricular units 2nd sem (grade)                   75
Curricular units 2nd sem (without evaluations)    153
Unemployment rate                                 136
Inflation rate                                    167
GDP                                                64
Target                                              0
dtype: int64

In [12]:
# Handling missing values with miceforest

# Pisahkan fitur dan target sebelum imputasi
X = df_mhs.drop('Target', axis=1)
y = df_mhs['Target']

# Membersihkan Nama Kolom Fitur
X.columns = [re.sub(r'[^A-Za-z0-9_]+', '_', col) for col in X.columns]

# Mengubah kolom yang bertipe data object menjadi category agar bisa dilakukan imputasi
object_cols = X.select_dtypes(include=['object']).columns

# Lakukan loop dan ubah tipe datanya
for col in object_cols:
    X[col] = X[col].astype('category')

# MICEforest akan menggunakan semua kolom di X untuk saling memprediksi nilai hilang
mhsk = mf.ImputationKernel(
  data=X,
  variable_schema=X.columns.to_list(),
  random_state=42,
)

# Jalankan proses imputasi
mhsk.mice(iterations=5, verbose=True)

# Dapatkan dataframe fitur yang sudah terisi penuh (imputed)
df_mhs_imputed = mhsk.complete_data()

print("\nDataframe Fitur Setelah Imputasi: ")
print(df_mhs_imputed)
print("\n")
df_mhs_imputed.info()
print("\n")

/usr/local/lib/python3.12/dist-packages/miceforest/ImputationKernel.py:370: UserWarning: [Marital_status,Nacionality] have very rare categories, it is a good idea to group these, or set the min_data_in_leaf parameter to prevent lightgbm from outputting 0.0 probabilities.
  warn(


Initialized logger with name mice 1-5
Dataset 0
1  | Curricular_units_2nd_sem_credited_ | Mother_s_occupation | GDP | Curricular_units_1st_sem_evaluations_ | Curricular_units_1st_sem_without_evaluations_ | Curricular_units_2nd_sem_grade_ | Daytime_evening_attendance_ | Marital_status | Previous_qualification_grade_ | Curricular_units_1st_sem_grade_ | Father_s_qualification | Curricular_units_2nd_sem_enrolled_ | International | Application_order | Debtor | Educational_special_needs | Displaced | Previous_qualification | Unemployment_rate | Admission_grade | Curricular_units_1st_sem_enrolled_ | Father_s_occupation | Mother_s_qualification | Curricular_units_2nd_sem_approved_ | Course | Curricular_units_2nd_sem_without_evaluations_ | Tuition_fees_up_to_date | Scholarship_holder | Inflation_rate | Application_mode | Curricular_units_1st_sem_approved_ | Nacionality | Gender | Age_at_enrollment | Curricular_units_2nd_sem_evaluations_ | Curricular_units_1st_sem_credited_
2  | Curricular_units

In [13]:
# Menampilkan dataset setelah diimputasi
df_mhs_imputed

Unnamed_0  Student_Id Marital_status  Application_mode  \
0             0        6634         single               1.0   
1             1        6459         single               1.0   
2             2        2238         single              42.0   
3             3        2479         single               1.0   
4             4        4260         single              44.0   
...         ...         ...            ...               ...   
3091       3091        6176         single              44.0   
3092       3092        8064         single               1.0   
3093       3093        2103         single               1.0   
3094       3094        8629         single              39.0   
3095       3095        6614         single              17.0   

      Application_order  Course Daytime_evening_attendance_  \
0                   3.0  9773.0                         day   
1                   1.0   171.0                         day   
2                   1.0  9500.0                         day   
3                   1.0  9070.0                         day   
4                   1.0  8014.0                         day   
...                 ...     ...                         ...   
3091                1.0  9085.0                         day   
3092                5.0  9238.0                         day   
3093                2.0  9500.0                         day   
3094                1.0  9500.0                         day   
3095                1.0  9670.0                         day   

      Previous_qualification  Previous_qualification_grade_ Nacionality  ...  \
0                        1.0                          137.0  Portuguese  ...   
1                        1.0                          136.0  Portuguese  ...   
2                        1.0                          120.0  Portuguese  ...   
3                        1.0                          141.0  Portuguese  ...   
4                       39.0                          120.0     Spanish  ...   
...                      ...                            ...         ...  ...   
3091                     1.0                          140.0  Portuguese  ...   
3092                     1.0                          126.0  Portuguese  ...   
3093                     1.0                          141.0  Portuguese  ...   
3094                     4.0                          150.0  Portuguese  ...   
3095                     1.0                          118.0  Portuguese  ...   

      Curricular_units_1st_sem_without_evaluations_  \
0                                               0.0   
1                                               0.0   
2                                               0.0   
3                                               0.0   
4                                               0.0   
...                                             ...   
3091                                            0.0   
3092                                            0.0   
3093                                            0.0   
3094                                            0.0   
3095                                            0.0   

      Curricular_units_2nd_sem_credited_  Curricular_units_2nd_sem_enrolled_  \
0                                    0.0                                 6.0   
1                                    0.0                                 0.0   
2                                    4.0                                 8.0   
3                                    0.0                                 6.0   
4                                    1.0                                 6.0   
...                                  ...                                 ...   
3091                                 1.0                                 6.0   
3092                                 0.0                                 6.0   
3093                                 0.0                                 8.0   
3094                                 2.0                  

In [14]:
# Agregasi lagi karena saat diimputasi tadi dipisahkan kolom target nya
df_mhs_new = pd.concat([df_mhs_imputed, train_labels], axis = 1)

df_mhs_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3096 entries, 0 to 3095
Data columns (total 39 columns):
 #   Column                                         Non-Null Count  Dtype   
---  ------                                         --------------  -----   
 0   Unnamed_0                                      3096 non-null   int64   
 1   Student_Id                                     3096 non-null   int64   
 2   Marital_status                                 3096 non-null   category
 3   Application_mode                               3096 non-null   float64 
 4   Application_order                              3096 non-null   float64 
 5   Course                                         3096 non-null   float64 
 6   Daytime_evening_attendance_                    3096 non-null   category
 7   Previous_qualification                         3096 non-null   float64 
 8   Previous_qualification_grade_                  3096 non-null   float64 
 9   Nacionality                              

In [15]:
# Encoding section (Menggunakan label encoder untuk kolom target saja, dan one hot encoder untuk beberapa kolom kategorikal yang independen)

# LABEL ENCODING
# Kolom yang akan diubah menggunakan One-Hot Encoding
ohe_cols = ['Marital_status', 'Daytime_evening_attendance_', 'Nacionality']
# Kolom yang akan diubah menggunakan Label Encoding
le_col = 'Target'

# Mengubah 'Graduate', 'Dropout', 'Enrolled' menjadi 0, 1, 2
le = LabelEncoder()

# Menerapkan fit_transform pada kolom target
df_mhs_new[le_col] = le.fit_transform(df_mhs_new[le_col])

# Membuat kamus penunjuk value pada kolom yang diubah menjadi angka apa
print(f"Mapping untuk kolom '{le_col}':")
for i, class_name in enumerate(le.classes_):
    print(f"  '{class_name}' -> {i}")
print("\n")

#ONE HOT ENCODER
# Menerapkan one hot encoder untuk beberapa kolom kategorikal
print("Menerapkan One-Hot Encoder pada Fitur")
df_mhs_encoded = pd.get_dummies(df_mhs_new, columns=ohe_cols,
                                prefix=ohe_cols, # Memberi nama kolom baru yang jelas
                                dummy_na=False,
                                dtype=int)

print("Dataframe Akhir (Setelah Encoding): ")
print(df_mhs_encoded.head())
print("\nTipe Data Kolom Setelah Encoding: ")
print(df_mhs_encoded.info())

Mapping untuk kolom 'Target':
  'Dropout' -> 0
  'Enrolled' -> 1
  'Graduate' -> 2


Menerapkan One-Hot Encoder pada Fitur
Dataframe Akhir (Setelah Encoding): 
   Unnamed_0  Student_Id  Application_mode  Application_order  Course  \
0          0        6634               1.0                3.0  9773.0   
1          1        6459               1.0                1.0   171.0   
2          2        2238              42.0                1.0  9500.0   
3          3        2479               1.0                1.0  9070.0   
4          4        4260              44.0                1.0  8014.0   

   Previous_qualification  Previous_qualification_grade_  \
0                     1.0                          137.0   
1                     1.0                          136.0   
2                     1.0                          120.0   
3                     1.0                          141.0   
4                    39.0                          120.0   

   Mother_s_qualification  Father_s_qual

In [16]:
# FINAL DATAFRAME SETELAH PREPROCESSING
df_mhs_encoded

Unnamed_0  Student_Id  Application_mode  Application_order  Course  \
0             0        6634               1.0                3.0  9773.0   
1             1        6459               1.0                1.0   171.0   
2             2        2238              42.0                1.0  9500.0   
3             3        2479               1.0                1.0  9070.0   
4             4        4260              44.0                1.0  8014.0   
...         ...         ...               ...                ...     ...   
3091       3091        6176              44.0                1.0  9085.0   
3092       3092        8064               1.0                5.0  9238.0   
3093       3093        2103               1.0                2.0  9500.0   
3094       3094        8629              39.0                1.0  9500.0   
3095       3095        6614              17.0                1.0  9670.0   

      Previous_qualification  Previous_qualification_grade_  \
0                        1.0                          137.0   
1                        1.0                          136.0   
2                        1.0                          120.0   
3                        1.0                          141.0   
4                       39.0                          120.0   
...                      ...                            ...   
3091                     1.0                          140.0   
3092                     1.0                          126.0   
3093                     1.0                          141.0   
3094                     4.0                          150.0   
3095                     1.0                          118.0   

      Mother_s_qualification  Father_s_qualification  Mother_s_occupation  \
0                        1.0                     1.0                  4.0   
1                       19.0                    19.0                  4.0   
2                        1.0                     2.0                  9.0   
3                        1.0                    19.0                  4.0   
4                        4.0                     4.0                  3.0   
...                      ...                     ...                  ...   
3091                     4.0                     4.0                  2.0   
3092                    38.0                     1.0                  9.0   
3093                    37.0                    37.0                  9.0   
3094                     3.0                    38.0                  3.0   
3095                     1.0                    19.0                  4.0   

      ...  Nacionality_Italian  Nacionality_Lithuanian  \
0     ...                    0                       0   
1     ...                    0                       0   
2     ...                    0                       0   
3     ...                    0                       0   
4     ...                    0                       0   
...   ...                  ...                     ...   
3091  ...                    0                       0   
3092  ...                    0                       0   
3093  ...                    0                       0   
3094  ...                    0                       0   
3095  ...                    0                       0   

      Nacionality_Moldova (Republic of)  Nacionality_Mozambican  \
0                                     0                       0   
1                                     0                       0   
2                                     0                       0   
3                                     0                       0   
4                                     0                       0   
...                                 ...                     ...   
3091                                  0                       0   
3092                                  0                       0   
3093                                  0                       0   
3094                                  0     